File: app/src/main/java/org/thoughtcrime/securesms/gcm/FcmUtil.java
package org.thoughtcrime.securesms.gcm;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.WorkerThread;

import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApp;
import com.google.firebase.messaging.FirebaseMessaging;

import org.signal.core.util.logging.Log;

import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public final class FcmUtil {

  private static final String TAG = Log.tag(FcmUtil.class);

  private static final long FCM_TOKEN_TIMEOUT_SECONDS = 5;

  /**
   * Retrieves the current FCM token. If one isn't available, it'll be generated.
   * Returns empty Optional if FCM is unavailable or times out.
   */
  @WorkerThread
  public static Optional<String> getToken(Context context) {
    String token = null;

    // Must be called manually if running outside of main process
    FirebaseApp.initializeApp(context);

    try {
      token = Tasks.await(FirebaseMessaging.getInstance().getToken(), FCM_TOKEN_TIMEOUT_SECONDS, TimeUnit.SECONDS);
    } catch (InterruptedException e) {
      Log.w(TAG, "Was interrupted while waiting for the token.");
    } catch (ExecutionException e) {
      Log.w(TAG, "Failed to get the token.", e.getCause());
    } catch (TimeoutException e) {
      Log.w(TAG, "FCM token request timed out after " + FCM_TOKEN_TIMEOUT_SECONDS + " seconds");
    }

    return Optional.ofNullable(TextUtils.isEmpty(token) ? null : token);
  }
}

File: demo/registration/src/main/java/org/signal/registration/sample/fcm/FcmUtil.kt
/*
 * Copyright 2025 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.signal.registration.sample.fcm

import android.content.Context
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.firebase.FirebaseApp
import com.google.firebase.messaging.FirebaseMessaging
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withTimeoutOrNull
import org.signal.core.util.logging.Log
import kotlin.time.Duration.Companion.seconds

/**
 * Utility functions for Firebase Cloud Messaging.
 */
object FcmUtil {

  private val TAG = Log.tag(FcmUtil::class)

  private val FCM_TOKEN_TIMEOUT = 5.seconds

  /**
   * Retrieves the FCM registration token if available.
   * Returns null if FCM is not available on this device.
   *
   * @param context Application context needed to initialize Firebase
   */
  suspend fun getToken(context: Context): String? {
    return try {
      FirebaseApp.initializeApp(context)

      val token = withTimeoutOrNull(FCM_TOKEN_TIMEOUT) {
        FirebaseMessaging.getInstance().token.await()
      }
      
      if (token != null) {
        Log.d(TAG, "FCM token retrieved successfully")
        token
      } else {
        Log.w(TAG, "FCM token request timed out after $FCM_TOKEN_TIMEOUT")
        null
      }
    } catch (e: Exception) {
      Log.w(TAG, "Failed to get FCM token", e)
      null
    }
  }

  /**
   * Checks if Google Play Services is available on this device.
   */
  fun isPlayServicesAvailable(context: Context): Boolean {
    val availability = GoogleApiAvailability.getInstance()
    val resultCode = availability.isGooglePlayServicesAvailable(context)
    return resultCode == ConnectionResult.SUCCESS
  }
}
